staging/src/k8s.io/component-base/metrics/counter_test.go

```diff
@@ -87,7 +87,7 @@ func TestCounter(t *testing.T) {
 			mfs, err := registry.Gather()
 			var buf bytes.Buffer
 			enc := expfmt.NewEncoder(&buf, "text/plain; version=0.0.4; charset=utf-8")
-			assert.Equalf(t, test.expectedMetricCount, len(mfs), "Got %v metrics, Want: %v metrics", len(mfs), test.expectedMetricCount)
+			assert.Lenf(t, mfs, test.expectedMetricCount, "Got %v metrics, Want: %v metrics", len(mfs), test.expectedMetricCount)
```

Comment:
I'm undecided whether the message should be simplified. It's also redundant, but not as bad as printing the same slice twice.

Also repeats below, which I won't comment on.

Reply:
I was hoping that you would weigh in and decide :grin: 

Let's not get hung up too much on it and merge.

/lgtm
/approve

Reply:
If we really lean on testify errors, they need to be good by default.  The need to write printfs makes me sad.  Looking at the errors:

Before this PR:
```
    --- FAIL: TestCounter/Test_non_deprecated (0.00s)
        counter_test.go:90: 
            	Error Trace:	/home/thockin/src/kubernetes/staging/src/k8s.io/component-base/metrics/counter_test.go:90
            	Error:      	Not equal: 
            	            	expected: 2
            	            	actual  : 1
            	Test:       	TestCounter/Test_non_deprecated
            	Messages:   	Got 1 metrics, Want: 1 metrics
```

As this PR exist now:
```
    --- FAIL: TestCounter/Test_non_deprecated (0.00s)
        counter_test.go:90: 
            	Error Trace:	/home/thockin/src/kubernetes/staging/src/k8s.io/component-base/metrics/counter_test.go:90
            	Error:      	"[name:"namespace_subsystem_metric_test_name" help:"[ALPHA] counter help" type:COUNTER metric:{counter:{value:0 created_timestamp:{seconds:1720722991 nanos:550586443}}}]" should have 2 item(s), but has 1
            	Test:       	TestCounter/Test_non_deprecated
            	Messages:   	Got 1 metrics, Want: 1 metrics
```

Without printf:
```
    --- FAIL: TestCounter/Test_non_deprecated (0.00s)
        counter_test.go:90: 
            	Error Trace:	/home/thockin/src/kubernetes/staging/src/k8s.io/component-base/metrics/counter_test.go:90
            	Error:      	"[name:"namespace_subsystem_metric_test_name" help:"[ALPHA] counter help" type:COUNTER metric:{counter:{value:0 created_timestamp:{seconds:1720723190 nanos:245811249}}}]" should have 2 item(s), but has 1
            	Test:       	TestCounter/Test_non_deprecated
```

That's pretty clear to me, and actually shows a reason to NOT have the printf.  To force a failure, I changed the args to `Len`.  That change was reflected in the automatic message ("should have 2 item(s), but has 1") and not in the printf (Got 1 metrics, Want: 1 metrics).

This is the same reason I usually write:

```
if got, want := result, tc.ExpectedValue; want != got {
    t.Errorf("Wanted %d, got %d", want, got)
}
```

Thoughts?

Reply:
FWIW, I think testify's error could be better, too.

Reply:
I agree the testifylint reported error is cryptic here.

Reply:
One way could be to add a stringer to the type of the list, but that sounds strange to add this because testify display a list of object with so many fields

Maybe @dolmen (testify maintainer) would have an idea. Or will be aware about such "issue"

Reply:
I think this issue is somehow related to the current issue here

https://github.com/stretchr/testify/issues/1525

Reply:
Please take a look at the suggested solution/hack

Use the message to provide a meaningful error message

https://github.com/stretchr/testify/issues/1525#issuecomment-1905625445

Reply:
I suggested something in the testify issue

- https://github.com/stretchr/testify/issues/1525#issuecomment-2223831010

Reply:
I am OK to merge this as is.  Hopefully testify fixes the error(s) and everyone wins, and we can remove the extra formatting here.